### ISSUE TITLE
Slow in big instalaltions, with alot of users

### ISSUE THREAD

**Issue by tomasbjerre** [tomasbjerre/simple-bitbucket-commit-checker#53]
Probably because of this:
https://github.com/tomasbjerre/simple-bitbucket-commit-checker/blob/master/src/main/java/se/bjurr/sbcc/SbccUserAdminServiceImpl.java

**Comment by fredrikand** [comment]
More details: https://github.com/tomasbjerre/simple-bitbucket-commit-checker/issues/54

**Comment by tomasbjerre** [comment]
This is hopefully fixed now in 2.5. Open issue again if problem still exists.

**Comment by fredrikand** [comment]
I can confirm that the performance issue remains in 2.5. Here is a profile from a push of 56 commits.

```
[162584ms] - DefaultHookService hook callback git-pre-receive for repository 146
  [162583ms] - com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory$1#preReceive
    [41ms] - com.atlassian.stash.internal.repository.ref.restriction.RestrictionEnforcer#onReceive
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [4ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [2ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
    [162542ms] - com.atlassian.stash.internal.hook.repository.PreReceiveRepositoryHookAdapter#onReceive
      [162540ms] - boolean com.atlassian.stash.internal.hook.repository.InternalRepositoryHookService.visitEnabledHooks(Repository,Class,RepositoryHookSettingsCallback)
        [1ms] - List com.atlassian.stash.internal.repository.RepositoryHookDao.findEnabledByKey(InternalRepository,Collection)
        [162539ms] - se.bjurr.sbcc.SbccPreReceiveRepositoryHook#onReceive
          [102ms] - Page com.atlassian.bitbucket.commit.CommitService.getCommitsBetween(CommitsBetweenRequest,PageRequest)
            [61ms] - /usr/bin/git rev-list --format=%H%x02%P%x02%aN%x02%aE%x02%at%n%B%n%x03END%x04 --ignore-missing --stdin --
            [35ms] - Map com.atlassian.stash.internal.content.IndexedCommitDao.getProperties(Iterable,Iterable)
            [3ms] - Map com.atlassian.stash.internal.repository.RepositoryActivityDao.countCommentsByCommit(int,Set)
            [3ms] - Map com.atlassian.stash.internal.user.InternalUserService.mapUsersByEmail(Set)
              [1ms] - InternalNormalUser com.atlassian.stash.internal.user.StashUserDao.loadUser(User)
```

**Comment by tomasbjerre** [comment]
So that would be ~2.9 seconds per commit. And if each commit has unique author email / name, they would each result in a lookup of that name / email. Making it 1.4s per lookup.

It is a bit faster right? you first posted that 20 commits took 3 minutes. And now 58 commits, in 2.7 minutes.

**Comment by fredrikand** [comment]
Must confess that the numbers I posted before were not as accurate as the numbers here.
This is the same profile for version 2.4. Yes, It is **much** quicker. 162584ms vs 299459ms. Same number of commits. Same user/email in all 67 commits.

```
$ git log | grep "This commit is created to test sbcc" | wc -l
67
```
approx 2.4s per commit.

Conf:
* Size check (15000kb) enabled
* Require Author Name and Author email found in Bitbucket user base.

```
  [299459ms] - com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory$1#preReceive
    [50ms] - com.atlassian.stash.internal.repository.ref.restriction.RestrictionEnforcer#onReceive
      [3ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [2ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [4ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
    [299409ms] - com.atlassian.stash.internal.hook.repository.PreReceiveRepositoryHookAdapter#onReceive
      [299408ms] - boolean com.atlassian.stash.internal.hook.repository.InternalRepositoryHookService.visitEnabledHooks(Repository,Class,RepositoryHookSettingsCallback)
        [1ms] - List com.atlassian.stash.internal.repository.RepositoryHookDao.findEnabledByKey(InternalRepository,Collection)
        [299407ms] - se.bjurr.sbcc.SbccPreReceiveRepositoryHook#onReceive
          [202ms] - Page com.atlassian.bitbucket.commit.CommitService.getCommitsBetween(CommitsBetweenRequest,PageRequest)
            [157ms] - /usr/bin/git rev-list --format=%H%x02%P%x02%aN%x02%aE%x02%at%n%B%n%x03END%x04 --ignore-missing --stdin --
            [37ms] - Map com.atlassian.stash.internal.content.IndexedCommitDao.getProperties(Iterable,Iterable)
            [3ms] - Map com.atlassian.stash.internal.repository.RepositoryActivityDao.countCommentsByCommit(int,Set)
            [3ms] - Map com.atlassian.stash.internal.user.InternalUserService.mapUsersByEmail(Set)
              [1ms] - InternalNormalUser com.atlassian.stash.internal.user.StashUserDao.loadUser(User)
          [131122ms] - Page com.atlassian.bitbucket.user.UserAdminService.findUsers(PageRequest)
            [685ms] - Set com.atlassian.stash.internal.user.StashUserDao.loadUsers(Iterable)
```

**Comment by tomasbjerre** [comment]
"Require Author Name" and "Require Author Email" both results in a lookup. If you uncheck one of them, do you get halv the lookup time then? And if you uncheck both, is it still very slow then?

**Comment by tomasbjerre** [comment]
Would be very interesting to debug this, you may give that a try. There is a [setup-atlassian-sdk.sh](https://github.com/tomasbjerre/simple-bitbucket-commit-checker/blob/master/setup-atlassian-sdk.sh) in the root of the repo that installs the Atlassian SDK. And then you can start the server with `atlas-debug`. Clone repo, run `./gradlew eclipse`, import in Eclipse, remote debug from Eclipse on port `5005`.

**Comment by fredrikand** [comment]
10 commits, both "Require Author Name" and "Require Author Email" unchecked
```
[24854ms] - DefaultHookService hook callback git-pre-receive for repository 146
  [24852ms] - com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory$1#preReceive
    [47ms] - com.atlassian.stash.internal.repository.ref.restriction.RestrictionEnforcer#onReceive
      [3ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [2ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [4ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
      [1ms] - Set com.atlassian.bitbucket.user.UserService.getUsersById(Set,boolean)
    [24805ms] - com.atlassian.stash.internal.hook.repository.PreReceiveRepositoryHookAdapter#onReceive
      [24804ms] - boolean com.atlassian.stash.internal.hook.repository.InternalRepositoryHookService.visitEnabledHooks(Repository,Class,RepositoryHookSettingsCallback)
        [24802ms] - se.bjurr.sbcc.SbccPreReceiveRepositoryHook#onReceive
          [175ms] - Page com.atlassian.bitbucket.commit.CommitService.getCommitsBetween(CommitsBetweenRequest,PageRequest)
            [158ms] - /usr/bin/git rev-list --format=%H%x02%P%x02%aN%x02%aE%x02%at%n%B%n%x03END%x04 --ignore-missing --stdin --
            [11ms] - Map com.atlassian.stash.internal.content.IndexedCommitDao.getProperties(Iterable,Iterable)
            [2ms] - Map com.atlassian.stash.internal.repository.RepositoryActivityDao.countCommentsByCommit(int,Set)
            [3ms] - Map com.atlassian.stash.internal.user.InternalUserService.mapUsersByEmail(Set)
              [1ms] - InternalNormalUser com.atlassian.stash.internal.user.StashUserDao.loadUser(User)
```

**Comment by fredrikand** [comment]
If I cleared the Check commit size (kb) option the delay is completely gone. I kept both "Require Author Name" and "Require Author Email" enabled.
149ms for 10 commits.

I would say it is no longer the author lookup that is causing the delay. It is the size check.

**Comment by tomasbjerre** [comment]
The author lookup is cached, 10 minutes. That may be why its that fast... but I hope not! =)
And yes, the file size can be made much faster!

**Comment by tomasbjerre** [comment]
I released 2.6 now. Try it and tell me how that works. I added caching for the filesizes.

**Comment by fredrikand** [comment]
Performance issues solved in 2.6!

Great thanks!

**Comment by tomasbjerre** [comment]
Nice, thanks for test effort! =)
